### PR TITLE
loam_multisense

### PR DESCRIPTION
**Pull request by gtinchev**
Could we tag it as v1 and rename it to loam_multisense?

Videos available here:
https://github.com/ipab-slmc/wiki/issues/61#issuecomment-231030485

Still to test 90+ rotations.

### CODE REVIEW

File: src/scanRegistration.cpp [hunk 1]
```diff
@@ -832,9 +832,9 @@ void imuHandler(const sensor_msgs::Imu::ConstPtr& imuIn)
     //imuAccuPitch += timeDiff * imuIn->angular_velocity.y;
     // imuAccuYaw += timeDiff * imuIn->angular_velocity.z;
     
-    imuRoll[imuPointerLast] = roll;
-    imuPitch[imuPointerLast] = -pitch;
-    imuYaw[imuPointerLast] = -yaw;
+    imuRoll[imuPointerLast] = yaw;
```
Comment (mauricefallon) [inline src/scanRegistration.cpp:918]: @gtinchev : 
I don't think this isn't correct but I have a very good idea about what is going on here.
I think you are very close to resolving this. Lets talk in the afternoon

**Comment by mauricefallon** [comment]
I'm going to leave this PR open until you've been able to get to the bottom of the IMU behaviour

I just had a look at the loam code and its rough! You have done well to make progress

File: src/scanRegistration.cpp [hunk 2]
```diff
@@ -46,12 +50,26 @@ sensor_msgs::PointCloud2 laserCloudLast2;
...
+
 int cloudSortInd[1200];
 int cloudNeighborPicked[1200];
 
 int imuPointerFront = 0;
 int imuPointerLast = -1;
-const int imuQueLength = 50;
+// GT changes 1/07/16 - IMU HZ
+// const int imuQueLength = 50;
+const int imuQueLength = 100;
```
Comment (mauricefallon) [inline src/scanRegistration.cpp:72]: keep this change in mind

File: src/scanRegistration.cpp [hunk 3]
```diff
@@ -384,25 +492,33 @@ void laserCloudHandler(const sensor_msgs::PointCloud2ConstPtr& laserCloudIn2)
...
+  float m = 0;
   for (int i = 5; i < cloudSize - 6; i++) {
+    // calculate difference with the neighbour
     float diffX = laserCloud->points[i + 1].x - laserCloud->points[i].x;
     float diffY = laserCloud->points[i + 1].y - laserCloud->points[i].y;
     float diffZ = laserCloud->points[i + 1].z - laserCloud->points[i].z;
     float diff = diffX * diffX + diffY * diffY + diffZ * diffZ;
 
+    // incremental average calculation of the diff
+    m = m + ((diff-m)/(cloudSize-12));
```
Comment (mauricefallon) [inline src/scanRegistration.cpp:523]: added but not used?

Comment (gtinchev) [inline src/scanRegistration.cpp:523]: This was my addition to calculate incremental average in order to see what is a "normal" value for the diff. Need to clean it up once I'm done with the IMU behaviour (a simple rotation was not enough for the IMU, still debugging).

**Comment by mauricefallon** [comment]
im going to merge this as it is partly working, but its impossible for me to analyse or suggest errors in a pull request this large.

its preferable to break them up into much small manageable units in future

**Pull request merged**